File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        List<Integer> inOrderList = new LinkedList<>();
        helper(root, inOrderList);
        
        boolean isBST = true;
        int prev = inOrderList.get(0);
        for (int i = 1; i < inOrderList.size(); i++) {
            if (inOrderList.get(i) <= prev) {
                isBST = false;
                break;  // No need to continue checking if it's already false
            }
            prev = inOrderList.get(i);
        }

        return isBST;
    }

    private void helper(TreeNode treeNode, List<Integer> inOrderList) {
        if (treeNode == null) return;
        helper(treeNode.left, inOrderList);
        inOrderList.add(treeNode.val);
        helper(treeNode.right, inOrderList);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def solve(self, root, mn_node, mx_node):
        if root is None:
            return True
        if mn_node and root.val >= mn_node.val:
            return False
        if mx_node and root.val <= mx_node.val:
            return False
        return (
            self.solve(root.left, root, mx_node)
            and
            self.solve(root.right, mn_node, root)
        )
    def isValidBST(self, root: TreeNode | None) -> bool:
        return self.solve(root, None, None)